### DISCUSSION TITLE
Poor performance after a few hours

### DISCUSSION THREAD

**Discussion by ZerGo0** [LazyVim/LazyVim#326] [Q&A]
Hey, I've recently switched to LazyVim, but I've noticed some performance issues. Neovim seems to become noticably slower after 4 hours or so. Code suggestions take longer to show up and the overall responsiveness is way worse to the point where navigating through a file becomes hard or even impossible. Closing neovim and starting it again fixes it, but it's kind of annoying to do that every few hours.

Has anyone else noticed something similar?

I'm on a "low" power notebook with a i7 7500U. I usually have like 6 buffers open. I don't really have any crazy customizations only a few keymaps and that's about it. I have used pretty much the exact same custom config with AstromVim before and didn't have any issues.

**Comment by folke** [comment]
I haven't noticed any slow-downs myself, but would be good if you could debug this more to see what might be causing this.

LSP servers tend to take up a lot of memory, so maybe that's the issue here? When things are slow, can you check your memory usage on your machine? Both for Neovim and any running LSP servers?

What OS are you on?

**Reply by ZerGo0** [reply]
> LSP servers tend to take up a lot of memory, so maybe that's the issue here? When things are slow, can you check your memory usage on your machine? Both for Neovim and any running LSP servers?

I just had it happen, the memory usage was at like 800MB, but the CPU usage was a a constant 25% (not much considering I have 4 cores). I've noticed that it showed a notification that it is initializing the LSP (tsserver in this case). I already had the file open for at least 10 minutes. Running `LspRestart` fixed it, but it was stuck at the initialization for at least 5 minutes.
> What OS are you on?

EndeavourOS with KDE

**Reply by folke** [reply]
Then tsserver is the issue. That lsp server has known performance issues. Nothing I can do about that

**Reply by ZerGo0** [reply]
> Then tsserver is the issue. That lsp server has known performance issues. Nothing I can do about that

It's currently slowing down to the point where actions like fuzzy searching take a few seconds to appear. The nvim process is at 2.9GB memory usage, `LspRestart` didn't fix it this time.

Any idea?

**Reply by folke** [reply]
What happens when you close some or al buffers? Does the memory usage drop?

**Reply by ZerGo0** [reply]
> What happens when you close some or al buffers? Does the memory usage drop?

No closing a buffer doesn't change the memory usage. I only had 2 buffers open (less than 100 lines Svelte files) and Neo-Tree.

**Reply by thallada** [reply]
Any ideas for how to get around this issue of tsserver slowing neovim down? I can't call it a serious editor if input lags to 1 second per character.

**Reply by viocost** [reply]
I am having performance issues with other servers, clangd for example, and the project isn't large at all.

**Reply by Demianeen** [reply]
@ZerGo0, I did make a PR for [vtsls extra](https://github.com/LazyVim/LazyVim/pull/2858). It is the lsp that vscode uses under the hood. It basically extracted from vscode. It uses less RAM and is generally faster.

**Reply by axelhj** [reply]
I'm having similar issues with built-in LSP and Neovim 10.0 (from master). tsserver and csharp_ls is particularly bad but it varies quite a bit.

**Reply by koplas** [reply]
> > What happens when you close some or al buffers? Does the memory usage drop?
> 
> No closing a buffer doesn't change the memory usage. I only had 2 buffers open (less than 100 lines Svelte files) and Neo-Tree.

I have the same issue with a svelte project and neo-tree.

**Reply by Demianeen** [reply]
Hey, have you tried https://github.com/Zeioth/garbage-day.nvim?

**Reply by kapral18** [reply]
garbage day is messing up with lsp and causes unnecessary crashes and doesn't play well with other plugin, at least that was with me that's why I removed it. 

In my personal experience nothing beats [typescript-tools](https://github.com/pmizio/typescript-tools.nvim). It's the only de-facto well working fast ts lsp, better than any other, including vtsls, especially for mid-to-large projects. The catch is you have to disable codelens and inlays to get the most out of it.

**Comment by jackielii** [comment]
It could be the lsp log? For me it grew to 10GB

**Reply by kapral18** [reply]
It was also affecting me so I `vim.lsp.set_log_level("off")` and only enable it when crashes happen to debug. And turn it back off.

**Comment by idelice** [comment]
It happens to me when I'm working on `tsx` files. It's _pretty_ bad like horrible.

I'm forced to switch to vscode most times..

**Reply by kapral18** [reply]
use https://github.com/pmizio/typescript-tools.nvim

**Comment by zaynOm** [comment]
I'm having the same issue when working with `tsx`  files as @idelice said. I at least conform that it's not from the lsp server, I restarted them all, the only thing that fix this is a nvim restart

I need to test it in different environment, but it would be better if we get some suggestion on how to profile it.

**Reply by koplas** [reply]
Some of the bad performance was because of the `lualine.nvim` configuration. Create a file under `~/.config/nvim/pulgins/lualine.lua` or your other neovim config folder and insert the following:

```lua
return {
  {
    "nvim-lualine/lualine.nvim",
    event = "VeryLazy",
    opts = function()
      return {
        --[[add your custom lualine config here]]
      }
    end,
  },
}
```

I suspect that displaying in what function/tag the cursor is located causes some performance problems. It still sometimes stutters, especially when saving the file.

**Reply by zaynOm** [reply]
I appreciate your response, I just removed the cursor location, let see if that fixes it

here is my `lualine.lua` config

update:
that does not fix the issue

```lua
return {
  "nvim-lualine/lualine.nvim",
  event = "VeryLazy",
  opts = {
    options = {
      section_separators = { left = "", right = "" },
      component_separators = { left = "|", right = "|" },
    },
    sections = {
      lualine_b = { "branch", "diff" },
      lualine_x = {
        -- stylua: ignore
        {
          function() return require("noice").api.status.command.get() end,
          cond = function() return package.loaded["noice"] and require("noice").api.status.command.has() end,
          color = function() return LazyVim.ui.fg("Statement") end,
        },
        -- stylua: ignore
        {
          function() return require("noice").api.status.mode.get() end,
          cond = function() return package.loaded["noice"] and require("noice").api.status.mode.has() end,
          color = function() return LazyVim.ui.fg("Constant") end,
        },
        -- stylua: ignore
        {
          function() return "  " .. require("dap").status() end,
          cond = function() return package.loaded["dap"] and require("dap").status() ~= "" end,
          color = function() return LazyVim.ui.fg("Debug") end,
        },
        -- stylua: ignore
        {
          require("lazy.status").updates,
          cond = require("lazy.status").has_updates,
          color = function() return LazyVim.ui.fg("Special") end,
        },
      },
      lualine_y = {}, -- removes cursor location
      lualine_z = {}, -- removes time
    },
  },
}
```

**Comment by kapral18** [comment]
This hits home with me, and I've been struggling with this for a long time. I think I've tried it all. 

Basically here is the recipe that works for me now:

1. Even though it's no longer actively maintained STILL the one and only unmatched https://github.com/pmizio/typescript-tools.nvim
2. This is my current config feel free to tweak https://github.com/kapral18/dotfiles/blob/52395e2935b122d291b3bf47e6bb64f3b4962d1d/home/dot_config/exact_nvim/exact_lua/exact_plugins/typescript.lua#L144-L163 but make sure you don't set `tsserver_max_memory` to auto because it's not auto, set to a specific number. Make sure to switch off codelens.
3. Switch off lsp logging `vim.lsp.set_log_level("off")` and switch it on only when needed for investigation.
4. Disable `inlay_hints` they have a performance cost, https://github.com/kapral18/dotfiles/blob/main/home/dot_config/exact_nvim/exact_lua/exact_plugins/lsp.lua#L20
5. Disable noice.nvim, it's an awesome plugin and brings a great look and feel but it's been a source of amazing amount of bugs and lags for a long time that I was attributing to other plugins.
6. Disable `lualine` sections and increase refresh timeouts. Here is mine: https://github.com/kapral18/dotfiles/blob/52395e2935b122d291b3bf47e6bb64f3b4962d1d/home/dot_config/exact_nvim/exact_lua/exact_plugins/lualine.lua
7. Disable  "lukas-reineke/indent-blankline.nvim" (comes with lazyvim) so https://github.com/kapral18/dotfiles/blob/52395e2935b122d291b3bf47e6bb64f3b4962d1d/home/dot_config/exact_nvim/exact_lua/exact_plugins/indent-blankline.lua
8. Set fast matchparen timeouts https://github.com/kapral18/dotfiles/blob/52395e2935b122d291b3bf47e6bb64f3b4962d1d/home/dot_config/exact_nvim/exact_lua/exact_config/options.lua#L66-L67
9. Syntax off, spell off, foldenable off https://github.com/kapral18/dotfiles/blob/52395e2935b122d291b3bf47e6bb64f3b4962d1d/home/dot_config/exact_nvim/exact_lua/exact_config/options.lua#L98-L102
10. Learn to profile using https://github.com/stevearc/profile.nvim
11. And debug using https://github.com/jbyuki/one-small-step-for-vimkind
12. From time to time `rm -rf ~/.cache/nvim ~/.local/share/nvim/ ~/.local/state/nvim`
13. When syntax lags happen, sometime it helps to reinstall tree-sitter that comes as a dependency of neovim. On mac it's `brew uninstall neovim && brew install neovim`. `brew reinstall neovim` didn't reinstall tree-sitter for me for some reason..
14. And finally, almost always works `sudo reboot now` :D

This should keep performance pretty manageable.

**Reply by laurensdc** [reply]
Do you have any updates now that we're a few years later?

**Reply by tronikelis** [reply]
Keep running into this with rust-analyzer lsp, my config is minimal, I haven't profiled this yet, but I know that rust-analyzer seems to leak memory and that makes neovim slow, but why is simple editing and navigating slow :thinking: you would think neovim lsp client is fully async and doesn't interrupt the user editing

**Comment by nemanjamalesija** [comment]
I struggled with this a lot. I tried the solutions above, and they made the situation more manageable, providing temporary relief, but they did not address the real issue. After a lot of research, I managed to pinpoint the cause.

The real problem was too many active buffers and LSP clients stacking. I only use fzf, Harpoon, and a file explorer to switch buffers and never properly close them. This caused LSP clients to accumulate per buffer, and after some time (1–2 hours of usage) my Neovim would start lagging badly. I limited the number of buffers to 10 to prevent this

```
vim.api.nvim_create_autocmd("BufAdd", {
    callback = function()
        local bufs = vim.fn.getbufinfo({ buflisted = 1 })
        if #bufs > 10 then
            for _, buf in ipairs(bufs) do
                if buf.bufnr ~= vim.api.nvim_get_current_buf() then
                    vim.cmd("bdelete " .. buf.bufnr)
                    break
                end
            end
        end
    end,
})
```

**Reply by thallada** [reply]
Interesting insight. I wonder if using a plugin like [hbac.nvim](https://github.com/axkirillov/hbac.nvim) or [nvim-early-retirement](https://github.com/chrisgrieser/nvim-early-retirement) would work better (and be less disruptive) for cleaning up unused buffers.

**Reply by nemanjamalesija** [reply]
@thallada I have considered that option, but I decided I don’t need another plugin to manage what is essentially a simple issue.

 I never work with more than 10 buffers at the same time, so I don’t need a way to smart manage them. I also thought plugin overhead is higher than just this simple piece of lua code. Why do you think a plugin would be better?

**Reply by thallada** [reply]
Makes sense for your use-case then. I personally don't keep track of what buffers I have opened and use oil.nvim to keep opening new files without closing the old buffer first so I tend to accumulate a ton of buffers in the background. I would prefer to just auto-close any buffer I don't currently have displayed in a split in a tab and not ever worry about manually managing buffers.

**Comment by nemanjamalesija** [comment]
Also, another update for everyone following this thread. I started monitoring my RAM usage over the last couple of weeks. LSP clients, particularly TypeScript and Copilot, leak memory. I cannot find an efficient way to prevent this. They consume more and more memory over time, and because I run a 16GB RAM Mac with a full development setup, I don't have much RAM to spare. Lags occur when LSP clients start consuming too much of it. In those cases, the only solution is to restart Neovim. It's annoying, but combined with the solution I provided above, it makes things more manageable.

**Reply by kapral18** [reply]
you are using raw ts_ls. There is really only one working lsp for typescript in neovim and that's tstools. Even abandoned it's still better than any alternative atm. Here is mine if it helps https://github.com/kapral18/dotfiles/blob/01717eabbf8d4ccce6be3d36c1877d61e9be9865/home/dot_config/exact_nvim/exact_lua/exact_plugins/typescript-tools.lua

**Reply by kapral18** [reply]
also check copilot.vim instead of copilot.lua it's not as fancy but it's not leaking

https://github.com/kapral18/dotfiles/blob/01717eabbf8d4ccce6be3d36c1877d61e9be9865/home/dot_config/exact_nvim/exact_lua/exact_plugins/ai.lua#L2-L10
https://github.com/kapral18/dotfiles/blob/01717eabbf8d4ccce6be3d36c1877d61e9be9865/home/dot_config/exact_nvim/exact_lua/exact_plugins/completion.lua#L72-L79

**Comment by ForgotenMail** [comment]
I have noticed a similar thing, I can often type things faster than lazyvim will register such as :w or moving between splits

**Comment by diostoup1522** [comment]
Is anyone else experiencing this too? For me it was vtsls eating up so much memory.

**Reply by tronikelis** [reply]
Yes, many people are experiencing this, please read the thread...
